marlowe-cli/src/Language/Marlowe/CLI/Test/CLI/Interpret.hs

```diff
@@ -0,0 +1,528 @@
+  let
+    curr = List.NonEmpty.head ciPlan
+  inputs <- for coInputs decodeInputJSON
+  minimumTime <- toPOSIXTime coMinimumTime
+  maximumTime <- toPOSIXTime coMaximumTime
+  era <- view eraL
+  logLabeledMsg co $ "Inputs: " <> show (pretty inputs)
+  new <- runLabeledCli era co $ prepareTransactionImpl
+    curr
+    inputs
+    minimumTime
+    maximumTime
+    True
+
+  let
+    new' = case coOverrideMarloweState of
+      Just customState -> new { mtState = customState }
+      Nothing          -> new
+    plan = new' <| ciPlan
+    marloweContract' = marloweContract{ ciPlan = plan }
+
+  modifying (contractsL . coerced) $ Map.insert coContractNickname marloweContract'
+
+interpret co@Publish {..} = do
+  logLabeledMsg co "Using reference scripts to the current Marlowe validator."
+  use publishedScriptsL >>= \case
+    Nothing -> do
+      logLabeledMsg co "Publishing the scripts."
+      marloweScriptRefs <- publishCurrentValidators coPublishPermanently coPublisher
+      assign publishedScriptsL (Just marloweScriptRefs)
+    Just _ -> do
+      throwLabeledError co "Scripts were already published during this test scenario."
+
+interpret AutoRun {..} = do
+  executionMode <- view executionModeL
+  case executionMode of
+    OnChainMode {} -> do
+      marloweContract@CLIContractInfo {..} <- findCLIContractInfo coContractNickname
+      let
+        plan = do
+          let
+            whole = reverse $ List.NonEmpty.toList ciPlan
+            threadLength = foldrMarloweThread (const (+ 1)) 0
+          case ciThread of
+            Just thread -> do
+              let
+                l = overAnyMarloweThread threadLength thread
+              drop l whole
+            Nothing -> whole
+        step mTh mt = do
+          let
+            prev :: Maybe (MarloweTransaction lang era, C.TxIn)
+            prev = do
+              pmt <- overAnyMarloweThread getCLIMarloweThreadTransaction <$> mTh
+              txIn <- overAnyMarloweThread marloweThreadTxIn =<< mTh
+              pure (pmt, txIn)
+            invalid = Just True == coInvalid
+
+          (txBody, mTxIn) <- autoRunTransaction ciCurrency ciSubmitter prev mt invalid
+          case (mTh, mTxIn) of
+            (Nothing, Nothing) -> throwError "[AutoRun] Creation of the Marlowe thread failed."
+            (Nothing, Just txIn) -> pure $ Just $ anyMarloweThreadCreated (mt, txBody) txIn
+            (Just th, _) -> case anyCLIMarloweThread (mt, txBody) mTxIn th of
+              Just th' -> pure $ Just th'
+              Nothing  -> throwError "[AutoRun] Extending of the Marlowe thread failed."
+
+      thread' <- foldM step ciThread plan
+      let
+        marloweContract' = marloweContract { ciThread = thread' }
+      modifying (contractsL . coerced)  $ Map.insert coContractNickname marloweContract'
+    SimulationMode -> do
+      -- TODO: We should be able to run balancing in here even in the simulation mode
```

Comment:
Yes, `Language.Marlowe.Runtime.Transaction.Safety` does this already.

Reply:
@bwbush I created a ticket for this (PLT-5583). May I resolve this for now and proceed and merge?